R/teal_slice.R

```diff
@@ -268,60 +275,50 @@ c.teal_slice <- function(...) {
   ans
 }
 
+#' @keywords internal
+#' @export
+as.list.teal_slice <- function(x) {
+  formals <- if (inherits(x, "teal_slice_expr")) {
+    formals(filter_expr)
+  } else {
+    formals(filter_var)
+  }
+
+  x <- if (shiny::isRunning()) {
+    shiny::reactiveValuesToList(x)
+  } else {
+    shiny::isolate(shiny::reactiveValuesToList(x))
+  }
+
+  formal_args <- setdiff(names(formals), "...")
+  extra_args <- setdiff(names(x), formal_args)
+
+  x[c(formal_args, extra_args)]
+}
+
 
-# format method for teal_slice
 #' @export
+#' @param show_all `logical(1)` should parameters set to NULL be returned
+#' @param nchar `integer(1)` or `NULL` indicating the number of character of the output width
 #' @rdname teal_slice
 #' @keywords internal
 #'
-format.teal_slice <- function(x, show_all = FALSE, ...) {
+format.teal_slice <- function(x, show_all = FALSE, nchar = 40, ...) {
   checkmate::assert_flag(show_all)
+  checkmate::assert_integerish(nchar, null.ok = TRUE)
 
-  x <- if (shiny::isRunning()) {
-    rev(shiny::reactiveValuesToList(x))
-  } else {
-    rev(shiny::isolate(shiny::reactiveValuesToList(x)))
-  }
+  x_list <- as.list(x)
+  if (!show_all) x_list <- Filter(Negate(is.null), x_list)
 
-  name_width <- max(nchar(names(x)))
-  format_value <- function(v) {
-    if (is.null(v)) {
-      return("NULL")
-    }
-    if (is.language(v)) {
-      v <- deparse1(v)
-    }
-    v <- paste(v, collapse = " ")
-    if (nchar(v) > 30L) {
-      v <- paste0(substr(v, 1, 26), "...")
-    }
-    v
-  }
-  ind <- intersect(names(x), names(formals(filter_var)))
-  xx <- x[ind]
-  hm <- "teal_slice"
-  for (i in seq_along(xx)) {
-    element_name <- format(names(xx)[i], width = name_width)
-    if (is.null(xx[[i]]) && !show_all) next
-    element_value <- format_value(xx[[i]])
-    hm <- append(hm, sprintf(" $ %s: %s", element_name, element_value))
-  }
+  x_json <- teal_slice_to_json(x_list)
+  x_json_c <- center_json(x_json)
 
-  ind <- setdiff(names(x), names(formals(filter_var)))
-  if (length(ind) != 0L) {
-    xx <- x[ind]
-    hm <- append(hm, " .. additional information")
-    for (i in seq_along(xx)) {
-      element_name <- format(names(xx)[i], width = name_width)
-      element_value <- format_value(xx[[i]])
-      hm <- append(hm, sprintf("     $ %s: %s", element_name, element_value))
-    }
-  }
-  paste(c(hm, ""), collapse = "\n")
+  if (!is.null(nchar)) x_json_c <- trim_json(x_json_c, nchar)
```

Comment:
`trim_character` in definition, `trim_json` in use

Reply:
weeeeeeird.